backend/crates/eiffelvis_core/src/graph.rs

```diff
@@ -0,0 +1,110 @@
+pub trait Key: PartialEq + Eq + Copy + Clone + std::fmt::Debug + std::hash::Hash + Send {}
+pub trait Idx: PartialEq + Eq + Copy + Clone + std::fmt::Debug + std::hash::Hash + Send {}
+impl<T> Idx for T where T: PartialEq + Eq + Copy + Clone + std::fmt::Debug + std::hash::Hash + Send {}
+
+/// Meta graph trait, defines the base types of a graph such as key and data types
+pub trait Meta {
+    type Idx: Idx;
+    type Key: Key;
+    type Data;
+    type EdgeData;
+}
+
+/// Main graph trait
+pub trait Graph:
+    Meta + ItemIter + Indexable<<Self as Meta>::Idx> + Indexable<<Self as Meta>::Key>
+{
+    /// Compares 2 indicies, this is not [Ord] as ordering may depend on graph internal state
+    fn cmp_index(&self, lhs: Self::Idx, rhs: Self::Idx) -> std::cmp::Ordering;
+
+    /// Returns the total amount of nodes this graph holds
+    fn node_count(&self) -> usize;
+
+    /// Creates a new edge with given data without any edges
+    fn add_node(&mut self, key: Self::Key, data: Self::Data) -> Option<Self::Idx>;
+
+    /// Creates a new edge between a and b with given edge data
+    fn add_edge(&mut self, a: Self::Key, b: Self::Key, data: Self::EdgeData);
+
+    /// Convience function, implemented in terms of [Graph::add_node] and [Graph::add_edge]
+    fn add_node_with_edges<I>(
+        &mut self,
+        key: Self::Key,
+        data: Self::Data,
+        edges: I,
+    ) -> Option<Self::Idx>
+    where
+        I: Iterator<Item = (Self::Key, Self::EdgeData)>,
+    {
+        let id = self.add_node(key, data);
+        if id.is_some() {
+            for (target, data) in edges {
+                self.add_edge(key, target, data);
+            }
+        }
+        id
```

Comment:
is this correct?

Reply:
We forward the result of self.add_node, looks a bit odd indeed but can't really remember a good way to make this more implicit

Reply:
We forward the result of self.add_node, looks a bit odd indeed but can't really remember a good way to make this more implicit